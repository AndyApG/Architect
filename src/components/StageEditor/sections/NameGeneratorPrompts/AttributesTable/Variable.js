import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { Field } from 'redux-form';
import cx from 'classnames';
import VariablePreview from './VariablePreview';
import VariableChooser from './VariableChooser';
import VariableEditor from './VariableEditor';
import { getVariablesForNodeType } from '../../../../../selectors/variableRegistry';

class Variable extends Component {
  static propTypes = {
    unusedVariables: PropTypes.array,
    label: PropTypes.string,
    nodeType: PropTypes.string,
    type: PropTypes.string,
    options: PropTypes.array,
    variable: PropTypes.string,
    isEditing: PropTypes.bool,
    onToggleEdit: PropTypes.func.isRequired,
    onDelete: PropTypes.func.isRequired,
    onChooseVariable: PropTypes.func.isRequired,
    onChange: PropTypes.func.isRequired,
    value: PropTypes.any,
  };

  static defaultProps = {
    label: '',
    type: null,
    nodeType: null,
    options: null,
    unusedVariables: [],
    isEditing: false,
    value: undefined,
    variable: null,
  };

  get isNew() {
    return !this.props.variable;
  }

  render() {
    const {
      unusedVariables,
      variable,
      value,
      label,
      type,
      options,
      nodeType,
      isEditing,
      onToggleEdit,
      onDelete,
      onChange,
      onChooseVariable,
    } = this.props;

    const variableClasses = cx(
      'attributes-table-variable',
      { 'attributes-table-variable--edit': isEditing },
      { 'attributes-table-variable--new': this.isNew },
    );

    return (
      <div
        className={variableClasses}
        onClick={onToggleEdit}
      >
        { !this.isNew &&
          <div className="attributes-table-variable__preview">
            <Field
              name={variable}
              variable={variable}
              label={label}
              component={VariablePreview}
              onDelete={onDelete}
            />
          </div>
        }

        <div className="attributes-table-variable__edit">
          <VariableChooser
            show={this.isNew}
            nodeType={nodeType}
            onChooseVariable={onChooseVariable}
            unusedVariables={unusedVariables}
          />
          <VariableEditor
            show={!this.isNew}
            value={value}
            variable={variable}
            type={type}
            label={label}
            onChange={onChange}
            options={options}
          />
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state, props) => {
  if (!props.variable) { return {}; }

  const variablesForNodeType = getVariablesForNodeType(state, props.nodeType);
  const variableMeta = variablesForNodeType[props.variable];

  return {
    label: variableMeta.label,
    type: variableMeta.type,
    options: variableMeta.options || null,
  };
};

export { Variable };

export default connect(mapStateToProps)(Variable);
