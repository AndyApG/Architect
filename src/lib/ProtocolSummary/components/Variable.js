import React, { useContext } from 'react';
import SummaryContext from './SummaryContext';
import DualLink from './DualLink';
import { getVariableName } from './helpers';

const Variable = ({
  id,
}) => {
  const {
    index,
  } = useContext(SummaryContext);

  return (
    <DualLink
      to={`#variable-${id}`}
      className="protocol-summary-variable"
    >
      {getVariableName(index, id)}
    </DualLink>
  );
};

export default Variable;
