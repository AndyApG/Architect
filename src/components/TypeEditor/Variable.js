import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import {
  Field,
  isDirty as isFieldDirty,
  FormSection,
  change,
  formValueSelector,
} from 'redux-form';
import Guidance from '../Guidance';
import { ValidatedField } from '../Form';
import * as Fields from '../../ui/components/Fields';
import * as ArchitectFields from '../Form/Fields';
import Validations from './Validations';
import Options from './Options';
import ExpandableItem from '../Items/ExpandableItem';

const VARIABLE_TYPES = [
  'text',
  'number',
  'datetime',
  'boolean',
  'ordinal',
  'categorical',
  'layout',
  'location',
];

const VARIABLE_TYPES_WITH_OPTIONS = [
  'ordinal',
  'categorical',
];

const Variable = ({
  fieldId,
  isDirty,
  form,
  variableType,
  resetOptions,
  ...rest
}) => (
  <ExpandableItem
    open={isDirty}
    preview={(
      <FormSection name={fieldId}>
        <h3 className="variable__preview-title">
          <Field
            name="name"
            component={({ input: { value } }) => value || 'undefined'}
          />
          &nbsp;:&nbsp;
          <em>
            <Field
              name="type"
              component={({ input: { value } }) => value || 'undefined'}
            />
          </em>
        </h3>
        <p className="variable__preview-description">
          <Field
            name="description"
            component={({ input: { value } }) => value}
          />
        </p>
      </FormSection>
    )}
    {...rest}
  >
    <FormSection name={fieldId}>
      <Guidance contentId="guidance.registry.type.variable">
        <div>
          <ValidatedField
            name="name"
            component={Fields.Text}
            label="Variable name"
            validation={{ required: true }}
          />
          <Field
            name="description"
            component={Fields.Text}
            label="Description"
          />
          <ValidatedField
            name="type"
            className="form-field-container"
            component={ArchitectFields.Select}
            label="Variable type"
            options={VARIABLE_TYPES}
            validation={{ required: true }}
            onChange={resetOptions}
          >
            <option value="">&mdash; Select variable type &mdash;</option>
          </ValidatedField>

          { VARIABLE_TYPES_WITH_OPTIONS.includes(variableType) &&
            <Options
              name="options"
              label="Options"
              meta={{ form }}
            />
          }

          <Validations
            name="validation"
            label="Validations"
            variableType={variableType}
            meta={{ form }}
          />
        </div>
      </Guidance>
    </FormSection>
  </ExpandableItem>
);

Variable.propTypes = {
  fieldId: PropTypes.string.isRequired,
  form: PropTypes.string.isRequired,
  isDirty: PropTypes.bool,
  resetOptions: PropTypes.func.isRequired,
  variableType: PropTypes.string,
};

Variable.defaultProps = {
  isDirty: false,
  variableType: null,
};

const mapStateToProps = (state, { form, fieldId }) => ({
  isDirty: isFieldDirty(form)(state, fieldId),
  variableType: formValueSelector(form)(state, `${fieldId}.type`),
});

const mapDispatchToProps = (dispatch, { form, fieldId }) => ({
  resetOptions: () => {
    dispatch(change(form, `${fieldId}.options`, []));
    dispatch(change(form, `${fieldId}.validation`, []));
  },
});

export { Variable };

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(Variable);
