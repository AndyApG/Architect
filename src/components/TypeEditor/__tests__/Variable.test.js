/* eslint-env jest */

import React from 'react';
import { shallow } from 'enzyme';
import { createStore } from 'redux';
import Variable from '../Variable';

const mockStore = () =>
  createStore(() => {});

const mockProps = {
  store: mockStore(),
  form: 'foo',
  fieldId: 'bar',
};

describe('<Variable />', () => {
  it('can render', () => {
    const subject = shallow(<Variable {...mockProps} />);
    expect(subject).toMatchSnapshot();
  });
});
