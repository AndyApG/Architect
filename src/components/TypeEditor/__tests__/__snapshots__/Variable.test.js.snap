// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Variable /> can render 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Connect(Variable)
    fieldId="bar"
    form="foo"
    store={
        Object {
            "dispatch": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
            Symbol(observable): [Function],
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "dispatch": [Function],
      "fieldId": "bar",
      "form": "foo",
      "isDirty": false,
      "store": Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      },
      "variableType": null,
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "dispatch": [Function],
        "fieldId": "bar",
        "form": "foo",
        "isDirty": false,
        "store": Object {
          "dispatch": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
          Symbol(observable): [Function],
        },
        "variableType": null,
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
